### PR TITLE
Consolidate local gradient aggregation code paths.

### PR DESCRIPTION
**Pull request by bfrieds**
# Testing
all run on 2 GPUs

tensorflow_keras_mnist.py with optimizer args: gradient aggregation=3, profile_frequency=1, profile_filename=pedl_profiling.log
tensorflow_mnist_estimator.py with optimizer args: gradient_aggregation=2

verified it ran to completion and both GPUs had the same validation loss/accuracy

### CODE REVIEW

File: horovod/common/gradient_aggregation.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        # This is going to be N data structure holding the aggregated gradient updates
+        # for parameter updates. N is the number of parameters.
+        self.gpu_shadow_vars = []
+
+        # Used to know when to allreduce and apply gradients. We allreduce when `self.counter`
+        # is equal to `self.aggregation_frequency`. We apply gradients when `self.counter` is
+        # equal to 0.
+        self.counter = None
+
+        self._sparse_as_dense = False
```
Comment (aaron276h) [inline horovod/common/gradient_aggregation.py]: Why are we setting this to False?

File: horovod/_keras/__init__.py [hunk 2]
```diff
@@ -79,242 +76,19 @@ def get_gradients(self, loss, params):
...
+                self._profile_helper.init_profile_vars(sess=tf.keras.backend.get_session(op_input_list=()))
+                self._agg_helper.init_aggregation_vars(
+                    self.grads,
+                    sess=tf.keras.backend.get_session(op_input_list=()),
+                )
+                with tf.control_dependencies([self._profile_helper.profile_start()]):
+                    allreduced_grads = self._agg_helper.compute_gradients(tuple(self.grads))
                 with tf.control_dependencies(allreduced_grads):
-                    comm_end = profile_end()
+                    comm_end = self._profile_helper.profile_end()
```
[the remark this answers is not in the document]
Comment (bfrieds) [reply horovod/_keras/__init__.py]: no, you _need_ the control_dependency because these are tf operaitons

File: horovod/common/gradient_aggregation.py (hunk 1 above)
Comment (bfrieds) [reply horovod/common/gradient_aggregation.py]: good catch
